scapy/layers/ppp.py

```diff
@@ -552,18 +552,151 @@ class PPP_LCP_Discard_Request(PPP_LCP):
                    IntField("magic_number", None),
                    StrLenField("data", "", length_from=lambda p:p.len-8)]
 
+### Password authentication protocol (RFC 1334)
+
+_PPP_paptypes = {1: "Authenticate-Request",
+                 2: "Authenticate-Ack",
+                 3: "Authenticate-Nak"}
+
+
+class PPP_PAP(Packet):
+    name = "PPP Password Authentication Protocol"
+    fields_desc = [ByteEnumField("code", 1, _PPP_paptypes),
+                   XByteField("id",0),
+                   ShortField("len", None),
+                   StrLenField("data","",length_from=lambda p:p.len-4)]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *args, **kargs):
+        if _pkt:
+            o = ord(_pkt[0])
+            if o == 1:
+                return PPP_PAP_Request
+            elif o in [2,3]:
+                return PPP_PAP_Response
+        return cls
```

Comment:
You might want to handle the case where a `code` has been provided in `kargs`. Maybe you could have a look at `LLTDAttribute.dispatch_hook()` in `scapy/layers/lltd.py`?